src/v/cluster/tm_stm.h

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/persisted_stm.h"
+#include "config/configuration.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/types.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "raft/consensus.h"
+#include "raft/errc.h"
+#include "raft/logger.h"
+#include "raft/state_machine.h"
+#include "raft/types.h"
+#include "storage/snapshot.h"
+#include "utils/expiring_promise.h"
+#include "utils/mutex.h"
+
+#include <absl/container/flat_hash_map.h>
+
+#include <compare>
+
+namespace cluster {
+
+struct tm_etag {
+    int64_t log_etag{0};
+    int64_t mem_etag{0};
+
+    tm_etag inc_mem() const {
+        return tm_etag{
+          .log_etag = this->log_etag, .mem_etag = this->mem_etag + 1};
+    }
+
+    tm_etag inc_log() const {
+        return tm_etag{.log_etag = this->log_etag + 1, .mem_etag = 0};
+    }
+
+    auto operator<=>(const tm_etag&) const = default;
+
+    friend std::ostream& operator<<(std::ostream&, const tm_etag&);
+};
+
+struct tm_transaction {
+    enum tx_status {
+        ongoing,
+        preparing,
+        prepared,
+        aborting,
+        finished,
+    };
+
+    struct tx_partition {
+        model::ntp ntp;
+        model::term_id etag;
+    };
+
+    // id of an application executing a transaction. in the early
+    // drafts of Kafka protocol transactional_id used to be named
+    // application_id
+    kafka::transactional_id id;
+    // another misnomer in fact producer_identity identifies a
+    // session of transactional_id'ed application as any given
+    // moment there maybe only one session per tx.id
+    model::producer_identity pid;
+    // tx_seq identifues a transactions within a session so a
+    // triple (transactional_id, producer_identity, tx_seq) uniquely
+    // identidies a transaction
+    model::tx_seq tx_seq;
+    tx_status status;
+    std::vector<tx_partition> partitions;
+    // version of tm_tx, used to perform conditional updates
+    tm_etag etag;
+
+    friend std::ostream& operator<<(std::ostream&, const tm_transaction&);
+};
+
+struct tm_snapshot {
+    model::offset offset;
+    std::vector<tm_transaction> transactions;
+};
+
+/**
+ * TODO: rename TM to TC
+ * TC stands for transaction coordinator. It's a state machine
```

Comment:
I'd be in favor of a full name for the class (e.g. `class txn_coordinator` or whatever). then if you're in a situation where you're having to type it out a lot you can selectively use `using tm_stm = txn_coordinator` in a context as an alias for shorter lines).